crates/worker/src/s3_client.rs

```diff
@@ -0,0 +1,146 @@
+use crate::commands::compile::CompilationFile;
+use crate::errors::S3Error;
+use aws_sdk_s3::presigning::{PresignedRequest, PresigningConfig};
+use aws_sdk_s3::types::Object;
+use aws_sdk_s3::Client;
+use aws_smithy_types::byte_stream::ByteStream;
+use std::io::Write;
+use std::path::Path;
+use tracing::{error, warn};
+
+#[derive(Clone)]
+pub struct S3Client {
+    pub client: Client,
+    pub bucket_name: String,
+}
+
+impl S3Client {
+    pub fn new(client: Client, bucket_name: &str) -> Self {
+        Self {
+            bucket_name: bucket_name.to_string(),
+            client,
+        }
+    }
+
+    pub async fn extract_files(&self, dir: &str) -> Result<Vec<CompilationFile>, S3Error> {
+        let objects = self.list_all_keys(dir).await?;
+
+        let mut files = vec![];
+        for object in objects {
+            let key = object.key().ok_or(S3Error::InvalidObjectError)?;
+            let expected_size = object.size.ok_or(S3Error::InvalidObjectError)?;
+
+            let mut contents = Vec::with_capacity(expected_size as usize);
```

Comment:
nitpick: it's better to use `usize::try_from`

Reply:
Well, size of our files will be restricted by system, so try_from could be unwrapped right away.